bfvmm/include/vmcs/vmcs_intel_x64.h

```diff
@@ -1917,20 +1829,1860 @@ constexpr const auto VMCS_HOST_IA32_PERF_GLOBAL_CTRL                      = 0x00
 // 32bit Control Fields
 // -----------------------------------------------------------------------------
 
-constexpr const auto VMCS_PIN_BASED_VM_EXECUTION_CONTROLS                      = 0x0000000000004000UL;
-constexpr const auto VMCS_PRIMARY_PROCESSOR_BASED_VM_EXECUTION_CONTROLS        = 0x0000000000004002UL;
-constexpr const auto VMCS_EXCEPTION_BITMAP                                     = 0x0000000000004004UL;
-constexpr const auto VMCS_PAGE_FAULT_ERROR_CODE_MASK                           = 0x0000000000004006UL;
-constexpr const auto VMCS_PAGE_FAULT_ERROR_CODE_MATCH                          = 0x0000000000004008UL;
-constexpr const auto VMCS_CR3_TARGET_COUNT                                     = 0x000000000000400AUL;
-constexpr const auto VMCS_VM_EXIT_CONTROLS                                     = 0x000000000000400CUL;
-constexpr const auto VMCS_VM_EXIT_MSR_STORE_COUNT                              = 0x000000000000400EUL;
-constexpr const auto VMCS_VM_EXIT_MSR_LOAD_COUNT                               = 0x0000000000004010UL;
-constexpr const auto VMCS_VM_ENTRY_CONTROLS                                    = 0x0000000000004012UL;
-constexpr const auto VMCS_VM_ENTRY_MSR_LOAD_COUNT                              = 0x0000000000004014UL;
-constexpr const auto VMCS_VM_ENTRY_INTERRUPTION_INFORMATION_FIELD              = 0x0000000000004016UL;
-constexpr const auto VMCS_VM_ENTRY_EXCEPTION_ERROR_CODE                        = 0x0000000000004018UL;
-constexpr const auto VMCS_VM_ENTRY_INSTRUCTION_LENGTH                          = 0x000000000000401AUL;
+namespace intel_x64
+{
+namespace vmcs
+{
+
+namespace pin_based_vm_execution_controls
+{
+    constexpr const auto addr = 0x0000000000004000UL;
+    constexpr const auto name = "pin_based_vm_execution_controls";
+
+    inline auto get()
+    { return vm::read(addr, name); }
+
+    template<class T> void set(T val)
+    { vm::write(addr, val, name); }
+
+    inline bool exists() noexcept
+    { return true; }
+
+    namespace external_interrupt_exiting
+    {
+        constexpr const auto mask = 0x00000001UL;
+        constexpr const auto from = 0;
+        constexpr const auto name = "external_interrupt_exiting";
+
+        inline bool
+        is_enabled()
+        { return (vm::read(addr, name) & mask) != 0; }
+
```

Comment:
I think if you have a addr_msr, the code will be cut / paste (literally it will be the same for every control) which too me is better so that if you have to make a change, it can be done quickly for all of them.